### PR TITLE
feat: in operator [#78]

### PR DESCRIPTION
**Pull request by jacekk**
I am not exactly sure what am I doing here ;) but I hope this is a good starting point. Please advise, what else can I improve in this.

P.S.
Sorry for the the flow ignore. I couldn't make it to work.

### CODE REVIEW

File: src/parsers/in.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import crap from './crap'
+
+import type { TupleNodeType, PrimitiveNodeType } from '../types'
+
+import StringParser from './primitives/string'
+import ObjectParser from './object'
+
+export default P.seq(
+  StringParser.trim(crap),
+  P.regexp(/\s*in\s*/).trim(crap),
```
Comment (kantord) [inline src/parsers/in.js]: `\s` is not required here, as it is handled by `.trim(crap)`

File: src/parsers/in.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import P from 'parsimmon'
+import crap from './crap'
+
+import type { TupleNodeType, PrimitiveNodeType } from '../types'
+
+import StringParser from './primitives/string'
+import ObjectParser from './object'
+
+export default P.seq(
+  StringParser.trim(crap),
```
Comment (kantord) [inline src/parsers/in.js]: This only works for string literals. It'd be much useful if the operator worked with any kind of value, some examples:

```
$myString in {} where $mystring = "Hello"
```

or 

```
["Hello", "asd"][0] in {}
```

File: src/parsers/in.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  P.regexp(/\s*in\s*/).trim(crap),
+  ObjectParser.trim(crap).map(
```
Comment (kantord) [inline src/parsers/in.js]: Here it'd be also better to accept more stuff. It could be a variable, or for example an array/list instead of an object. For now it's enough to give runtime errors here, it's not necessary to enforce type in the parser here.

File: src/parsers/in.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    // $FlowFixMe
+    (found: mixed): [TupleNodeType] => found.value[0].value
```
Comment (kantord) [inline src/parsers/in.js]: This flow issue will go away automatically if you make the other changes :-)

File: src/parsers/in.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  )
+)
+  .map(([key, _, props]: [PrimitiveNodeType, mixed, [TupleNodeType]]): boolean => {
```
Comment (kantord) [inline src/parsers/in.js]: It's not required to execute code in the parser. Instead we generate an AST. (See `NodeType` for it's format).

We have separate generators that convert AST into JavaScript, and they reside in `./src/generators/`.

File: src/parsers/__tests__/in.test.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
+// @flow
+
+import parser from '../in'
+
+const trueCases = [
+  '"foo" in {"foo": "bar"}',
```
Comment (kantord) [inline src/parsers/__tests__/in.test.js]: You don't have to actually test for the values these will be evaluated to here. This file should merely test the parser.

Generators have separate test, which test for the code that is generated. Testing for different situations that don't have a syntactical difference should be handled there.

In this very case, actually the calculation of the result should be simply handled by the generated JavaScript program itself, so it's enough to test this with an integration test. For now, integration tests reside here:
https://github.com/kantord/emuto/blob/master/src/__tests__/interpreter.test.js#L326

This will also add some snapshot tests automatically, but the main point is testing the result.

Many time I implement the integration test first, and then implement the parser, then the generator, or anything else that is required. Usually I use the pattern mode in jest to only execute tests for the part that I'm working on, so that other failures don't annoy me. Once the integration test passes, that tells me that the feature is wholly implemented.

In this file it's enough to test an example of of the syntax (and dangerous edge cases) and an example of the resulting syntax tree. Here's an example: https://github.com/kantord/emuto/blob/master/src/parsers/__tests__/parentheses.test.js I believe the tests for this operator will not get any more complex than that of parentheses.

**Comment by jacekk** [comment]
@kantord I am really sorry, but the abstraction behind this project seems to be above me. After a few evenings of trial&error, I just give up. I don't want to block this feature to be implemented. That's for your effort put into all above tips.
